pkg/agent/proxy/proxier.go

```diff
@@ -868,6 +875,9 @@ func (p *proxier) OnServiceUpdate(oldService, service *corev1.Service) {
 	if oldService != nil && service != nil {
 		klog.V(2).InfoS("Processing Service UPDATE event", "Service", klog.KObj(service))
 	}
+	if !p.serviceLabelSelector.Matches(labels.Set(service.Labels)) {
+		return
+	}
```

Comment:
question: is this enough to filter out the services? could the services be present in `serviceMap` when `installServices` is called?

Reply:
- To filter out headless Service, add a selector `labels.NewRequirement(corev1.IsHeadlessService, selection.DoesNotExist, nil)` to  `serviceLabelSelector`; 
- when serviceProxyName is not set, add a selector `labels.NewRequirement(labelServiceProxyName, selection.DoesNotExist, nil)` to filter out the Services with the label;  
- when serviceProxyName is set, add a selector ` labels.NewRequirement(labelServiceProxyName, selection.DoubleEquals, []string{serviceProxyName})` to filter out other Services.

The Services filtered out won't be presented in `serviceMap` since they will not be presented in `serviceChanges` in event handler function, then in https://github.com/antrea-io/antrea/blob/a7062042ec7881a28d61a934b2d35c10d8e22282/pkg/agent/proxy/proxier.go#L768, they will not be presented in `serviceMap`.

Reply:
I took at the code. It seems deleting such service will lead to an Error log since it's never added to serviceMap but could be deleted from it. Can you check?
```
klog.Errorf("Service port %q doesn't exists", svcPortName)
```

Besides, I found there is already a function `ShouldSkipService` used to skip services, perhaps the new logic should be added there to be consistent and to avoid the above issue?

Reply:
@hongliangl I don't think you can use `labels.NewRequirement(corev1.IsHeadlessService, selection.DoesNotExist, nil)` to filter the headless Service. The well-known label is added by Controller to Endpoint instead of the Service. I think I pasted the comment of the label before. Please double check and refine this part.
```
	// IsHeadlessService is added by Controller to an Endpoint denoting if its parent
	// Service is Headless. The existence of this label can be used further by other
	// controllers and kube-proxy to check if the Endpoint objects should be replicated when
	// using Headless Services
	IsHeadlessService = "service.kubernetes.io/headless"
```

Reply:
> I took at the code. It seems deleting such service will lead to an Error log since it's never added to serviceMap but could be deleted from it. Can you check?
> 
> ```
> klog.Errorf("Service port %q doesn't exists", svcPortName)
> ```
> 
> Besides, I found there is already a function `ShouldSkipService` used to skip services, perhaps the new logic should be added there to be consistent and to avoid the above issue?

I moved the labelSelector to the same code block where `ShouldSkipService` is called. Then there is no error for the following cases:

- Add a Service with label `service.kubernetes.io/service-proxy-name`, then delete it,  AntreaProxy doesn't add them to `serviceMap` in `ADD` and `DELETE` event.
- Add a Service with label `service.kubernetes.io/service-proxy-name`, then update the Service by removing the label, AntreaProxy will sync the Service like syncing a new Service.
- Add a Service without label `service.kubernetes.io/service-proxy-name`, then update the Service by adding the lable, AntreaProxy will remove the Service like remove a stale Service.

Reply:
> @hongliangl I don't think you can use `labels.NewRequirement(corev1.IsHeadlessService, selection.DoesNotExist, nil)` to filter the headless Service. The well-known label is added by Controller to Endpoint instead of the Service. I think I pasted the comment of the label before. Please double check and refine this part.
> 
> ```
> 	// IsHeadlessService is added by Controller to an Endpoint denoting if its parent
> 	// Service is Headless. The existence of this label can be used further by other
> 	// controllers and kube-proxy to check if the Endpoint objects should be replicated when
> 	// using Headless Services
> 	IsHeadlessService = "service.kubernetes.io/headless"
> ```

@luolanzone You are right, the label is only added to Endpoints of headless Services automatically by Controller, rather than headless Services themselves. However, I observed that kube-proxy incorporates this logic. I supposed that this is used to skip some Services with labeling the well-known label service.kubernetes.io/headless manually in kube-proxy, although these Services may not real headless Services.

Reply:
kube-proxy adds the selector to avoid retriving endpoints of headless services, instead services. The selector is shared for all queries. No service will have that label.

Reply:
But in sig-network e2e test, the test will update the Service with label. See this link https://github.com/kubernetes/kubernetes/blob/027ac5a426a261ba6b66a40e79e123e75e9baf5b/test/e2e/network/service.go#L2281. I am not sure if we need to pass the test,  or the test is not reasonable, then we don't need to pass the test.

Reply:
The test wants to verify the service is not served but needs a way to check reachability, so it creates a fake headless service by leveraging the fact that service's labels will be copied to endpoints' labels. And the essential usage of the label is to prevent proxy from handling headless services' endpoints. I think you should skip Endpoints when it contains such label, instead of checking service's labels. Then the test should pass.